### PR TITLE
Update react deps

### PR DESCRIPTION
**Pull request by andrewballantyne**
Addresses: #591 

## Description

Upgrades dependencies for React and React Router & handles the necessary backwards breaking changes.

This is needed for the SDK as well. So we'll want to merge this into `f/dsg` once it goes into `main`.

## How Has This Been Tested?

Viewed each page, tested various functionalities for routes.

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: frontend/package.json [hunk 1]
```diff
@@ -52,12 +52,12 @@
     "lodash-es": "^4.17.15",
     "node-gyp": "^7.1.2",
     "node-sass": "^4.14.1",
-    "react": "^16.13.1",
+    "react": "^17.0.2",
```
Comment (andrewballantyne) [inline frontend/package.json:55]: Wanted 18... PF is not ready.
https://github.com/patternfly/patternfly-react/issues/7142

File: frontend/src/pages/notebookController/screens/server/NotebookServerRoutes.tsx [hunk 2]
```diff
@@ -1,27 +1,17 @@
...
-      </Route>
-      <Route>
-        <Redirect to={path} />
-      </Route>
-    </Switch>
+    <Routes>
+      <Route path="/" element={<NotebookServer />} />
+      <Route path="/spawner" element={<SpawnerPage />} />
+      <Route path="/:username/:type" element={<NotebookControlPanelRedirect />} />
+      <Route path="*" element={<Navigate to="/notebookController" replace />} />
```
Comment (andrewballantyne) [inline frontend/src/pages/notebookController/screens/server/NotebookServerRoutes.tsx:13]: We'll need to probably get constants involved here - FYI @dlabaj 

I didn't see an easy way to just reference "the path for this component"

File: frontend/src/utilities/useTrackHistory.ts [hunk 3]
```diff
@@ -1,20 +1,12 @@
...
-      const { pathname: nextPathname } = history.location;
-      if (pathname !== nextPathname) {
-        pathname = nextPathname;
-        fireTrackingEvent('page');
-      }
-    });
-    return () => unlisten();
-  }, [history]);
+    fireTrackingEvent('page'); // TODO: do we want to track WHICH page?
+  }, [pathname]);
```
Comment (andrewballantyne) [inline frontend/src/utilities/useTrackHistory.ts:11]: Weird hook -- it mutated `history.location` 🤔 and it doesn't track the route

Not sure if I got this right -- but the logic was weird.

File: frontend/src/pages/notebookController/screens/server/NotebookServerRoutes.tsx (hunk 2 above)
Comment (andrewballantyne) [reply frontend/src/pages/notebookController/screens/server/NotebookServerRoutes.tsx]: nvm, Juntao figured out a solution using `.`.

**Review by DaoDaoNoCode — APPROVED** [review]
/lgtm

**Pull request merged**